### PR TITLE
fix: make rate limiting auth-aware under shared proxy ips

### LINKED ISSUE #411: Bug: Authenticated users can be rate-limited by shared proxy IPs
## Summary
Today, API rate limiting blocks a request when either the IP bucket or the key bucket is exhausted.
In proxied/shared-egress environments, this can deny legitimate authenticated users even when their own token/user activity is reasonable. (Ref: #349, #390)

This issue proposes an auth-aware limiter:
- For authenticated requests: enforce user-based limits first, with IP as fallback/safety guard.
- For anonymous requests: keep IP-first behavior.

It also proposes changing forwarded-header fallback precedence (when trusted) from `x-real-ip` first to `x-forwarded-for` first hop.

### 1) Current deny condition is `ip OR key`
`convex/lib/httpRateLimit.ts:35`:

```ts
if (!ipResult.allowed || (keyResult && !keyResult.allowed)) {
```

This means an authenticated request is denied if the IP bucket (which may be a shared IP) has hit limits, even if token/key usage is healthy.

### 2) IP extraction favors Cloudflare header, then forwarded fallback
`convex/lib/httpRateLimit.ts:55-67`:
- `cf-connecting-ip` is used first.
- Fallback to forwarded headers is gated by `TRUST_FORWARDED_IPS=true`.
- Current fallback order is `x-real-ip` -> `x-forwarded-for` -> `fly-client-ip`.

- `X-Forwarded-For` is the common proxy-chain header and is more likely to be the right one for the end user. (See Refs)

### 3) App-domain API traffic is proxied to Convex
`vercel.json:4-5` rewrites:
- `/api/:path*` -> `https://<deployment>.convex.site/api/:path*`

`docs/deploy.md:44-50` documents the same path.

Live header sample from `https://clawhub.ai/api/v1/search?q=test` (2026-02-18) showed:
- `server: Vercel`
- `convex-usher: usher`

This confirms app-domain API requests traverse proxy layers before Convex, potentially resulting in shared IPs that are outside the user's control.

### 4) CLI defaults also target app-domain API
- `packages/clawdhub/src/cli/registry.ts:6` default registry is `https://clawhub.ai`.
- `public/.well-known/clawhub.json:2` also points `apiBase` to `https://clawhub.ai`.

So many CLI requests also enter via the app-domain proxy path by default.

## Why this causes false positives
- Shared NAT/proxy egress (office, cloud workers, CI runners, mobile carrier NAT) can concentrate many users behind one observed IP.
- With current `ip OR key` deny semantics, authenticated users inherit each other’s IP pressure.
- This is amplified when API calls pass through shared edge/proxy infrastructure.

### 1) Auth-aware limiter semantics
- If request has a valid API token user:
  - Enforce `user:<userId>` bucket (primary).
  - Keep IP bucket as fallback/safety guard only (for obvious abuse spikes), not as equal hard blocker.
- If request is anonymous:
  - Keep IP-based enforcement as today.

Implementation note:
- Reuse `getOptionalApiTokenUserId` from `convex/lib/apiTokenAuth.ts:30`.
- Build limiter keys as:
  - `user:<userId>` for authenticated requests
  - `ip:<ip>` for anonymous/fallback

### 2) Forwarded fallback precedence (when trusted)
- Keep `cf-connecting-ip` as first choice.
- When `TRUST_FORWARDED_IPS=true` and CF header is missing:
  - Prefer `x-forwarded-for` first hop
  - Then `x-real-ip`
  - Then `fly-client-ip`

Rationale:
- `X-Forwarded-For` is the common proxy-chain header and carries client + hop lineage.
- `X-Real-IP` is non-standard and single-hop, often proxy-specific.

### 3) Observability + headers
- Add internal structured logging for limiter decisions (`auth`, `user_allowed`, `ip_allowed`, `ip_source`).
- Continue returning standard rate-limit headers already in place.

## References
- RFC 7239 (`Forwarded`): https://datatracker.ietf.org/doc/html/rfc7239
- MDN `X-Forwarded-For`: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For
- NGINX real IP module (`X-Real-IP` / `X-Forwarded-For`): https://nginx.org/en/docs/http/ngx_http_realip_module.html
- Vercel request headers (`x-forwarded-for`, `x-real-ip`): https://vercel.com/docs/headers/request-headers.rsc
- Cloudflare original visitor IP (`CF-Connecting-IP`): https://developers.cloudflare.com/support/troubleshooting/restoring-visitor-ips/restoring-original-visitor-ips/

### PR DESCRIPTION
**Pull request by lc0rp**
## Summary

Today, API rate limiting blocks based on IP or key. In proxied/shared-egress environments, IP sharing can deny legitimate authenticated users. 

Ref issues: https://github.com/openclaw/clawhub/issues/349, https://github.com/openclaw/clawhub/issues/390

- Make API rate limiting prefer key for authenticated requests and keep IP enforcement for anonymous/invalid-token requests
- Invert trusted fallback precedence to `x-forwarded-for` -> `x-real-ip` -> `fly-client-ip`
- Add structured deny logs (`auth`, `userAllowed`, `ipAllowed`, `ipSource`, `hasClientIp`)
- Update HTTP API docs to reflect auth-aware limiter semantics

## Rationale
- current `ip OR key` deny behavior can block valid authenticated users behind shared egress/proxy IPs
- user-first enforcement for authenticated traffic reduces false positives while preserving anonymous IP controls
- `x-forwarded-for` first hop is the more common client-IP signal in trusted proxy chains

## Work done
- `convex/lib/httpRateLimit.ts`
  - switched to auth-aware decision model
  - switched trusted fallback precedence to `x-forwarded-for` first
  - added structured rate-limit deny logging
- `convex/lib/httpRateLimit.test.ts`
  - added auth-vs-anon enforcement regression coverage
  - added forwarded-header precedence coverage
- `convex/httpApiV1.handlers.test.ts`
  - adjusted auth mocks for stars routes under new limiter path
- `docs/http-api.md`, `docs/api.md`
  - documented auth-aware enforcement semantics

## Testing
```bash
bun run lint
bunx vitest run convex/lib/httpRateLimit.test.ts convex/httpApiV1.handlers.test.ts
bun run test
```

Checklist
- [x] Added test coverage and tested
- [x] bun run lint
- [x] Tested and working locally

Tag-teamed w/ codex 5.3 on this 🤖 

Fixes #411, #349, #390

<h3>Greptile Summary</h3>

Modified rate limiting from IP OR key enforcement to auth-aware enforcement: authenticated requests now consume only user buckets, while anonymous requests use IP buckets. This prevents legitimate authenticated users from being blocked by shared proxy IPs while preserving IP-based protection for anonymous traffic.

Key changes:
- Switched IP header precedence to `x-forwarded-for` → `x-real-ip` → `fly-client-ip` (more common proxy pattern)
- Authenticated requests skip IP bucket consumption entirely (addressed previous reviewer's concern)
- Added structured deny logging with auth status, IP source, and bucket states
- Fixed double header parsing in client SDK (addressed previous reviewer's concern)
- Enhanced client retry logic to respect `Retry-After` headers with jitter
- Updated rate limit headers to include both legacy (`X-RateLimit-*`) and standardized (`RateLimit-*`) formats
- Comprehensive test coverage for auth vs anonymous enforcement, token validation fallback, and header precedence

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge with no blocking issues
- The implementation correctly addresses the shared proxy IP problem with clean auth-aware logic, comprehensive test coverage validates all critical paths including edge cases, previous review concerns about IP consumption and header parsing have been properly fixed, documentation is thorough, and the changes follow established patterns in the codebase
- No files require special attention

<sub>Last reviewed commit: a176658</sub>

### CODE REVIEW

File: packages/clawdhub/src/http.ts [hunk 1]
```diff
@@ -171,14 +205,153 @@ async function readResponseTextSafe(response: Response): Promise<string> {
...
+
+function throwHttpStatusError(status: number, text: string, headers?: HeaderSource): never {
+  const message = buildHttpErrorMessage(status, text, headers)
+  const rateLimit = parseRateLimitInfo(headers)
   if (status === 429 || status >= 500) {
-    throw new Error(message)
+    throw new HttpStatusError(status, message, rateLimit)
   }
   throw new AbortError(message)
 }
```
[the remark this answers is not in the document]
Comment (lc0rp) [reply packages/clawdhub/src/http.ts:256]: Fixed in commit [661f019](https://github.com/openclaw/clawhub/pull/412/commits/661f0190b5ebdc3cedca089a5dac4eaf8551215c)

File: convex/lib/httpRateLimit.ts [hunk 2]
```diff
@@ -22,17 +22,28 @@ export async function applyRateLimit(
...
 ): Promise<{ ok: true; headers: HeadersInit } | { ok: false; response: Response }> {
+  const userId = await getOptionalApiTokenUserId(ctx, request)
   const ip = getClientIp(request) ?? 'unknown'
   const ipResult = await checkRateLimit(ctx, `ip:${ip}`, RATE_LIMITS[kind].ip)
-  const token = parseBearerToken(request)
-  const keyResult = token
-    ? await checkRateLimit(ctx, `key:${await hashToken(token)}`, RATE_LIMITS[kind].key)
+  const userResult = userId
+    ? await checkRateLimit(ctx, `user:${userId}`, RATE_LIMITS[kind].key)
     : null
```
[the remark this answers is not in the document]
Comment (lc0rp) [reply convex/lib/httpRateLimit.ts]: Fixed in [a176658](https://github.com/openclaw/clawhub/pull/412/commits/a176658d32abc6c7b789fbe3a1f46bb4e36f50b8)

**Comment by steipete** [comment]
Landed via temp rebase onto main.

- Gate: bun run lint && bun run build && bun run test
- Land commit: a176658d32abc6c7b789fbe3a1f46bb4e36f50b8
- Merge commit: da4469e1e084811f6a0c6708c20daf17e05b2297

Thanks @lc0rp!
